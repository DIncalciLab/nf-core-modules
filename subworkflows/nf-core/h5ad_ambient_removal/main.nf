include { CELDA_DECONTX               } from '../../../../modules/nf-core/celda/decontx'
include { CELLBENDER_REMOVEBACKGROUND } from '../../../../modules/nf-core/cellbender/removebackground'
include { CELLBENDER_MERGE            } from '../../../../modules/nf-core/cellbender/merge'
include { SOUPX                       } from '../../../../modules/nf-core/soupx'
include { SCVITOOLS_SCAR              } from '../../../../modules/nf-core/scvitools/scar'

workflow H5AD_AMBIENT_REMOVAL {
    take:
    ch_filtered_unfiltered
    tool

    main:
    ch_versions = Channel.empty()

    if (tool == 'none') {
        println "AMBIENT_RNA_REMOVAL: Not performed since 'none' selected."
        ch_h5ad = ch_filtered_unfiltered.map{ meta, filtered, unfiltered -> [meta, filtered] }
    }
    else if (tool == 'decontx') {
        CELDA_DECONTX(ch_filtered_unfiltered)
        ch_h5ad = CELDA_DECONTX.out.h5ad
        ch_versions = ch_versions.mix(CELDA_DECONTX.out.versions)
    }
    else if (tool == 'cellbender') {
        CELLBENDER_REMOVEBACKGROUND(ch_filtered_unfiltered.map{meta, filtered, unfiltered -> [meta, unfiltered]})
        ch_versions = ch_versions.mix(CELLBENDER_REMOVEBACKGROUND.out.versions)

        CELLBENDER_MERGE(ch_filtered_unfiltered.map{ meta, filtered, raw -> [meta.id, meta, filtered, raw] }
            .join(CELLBENDER_REMOVEBACKGROUND.out.h5.map{ meta, h5 -> [meta.id, h5] }, by: 0, failOnMismatch: true)
            .map{ id, meta, filtered, raw, h5 -> [meta, filtered, raw, h5] })
        ch_h5ad = CELLBENDER_MERGE.out.h5ad
        ch_versions = ch_versions.mix(CELLBENDER_MERGE.out.versions)
    }
    else if (tool == 'soupx') {
        SOUPX(ch_filtered_unfiltered)
        ch_h5ad = SOUPX.out.h5ad
        ch_versions = ch_versions.mix(SOUPX.out.versions)
    }
    else if (tool == 'scar') {
        SCVITOOLS_SCAR(ch_filtered_unfiltered)
        ch_h5ad = SCVITOOLS_SCAR.out.h5ad
        ch_versions = SCVITOOLS_SCAR.out.versions
    }
    else {
        error "AMBIENT_REMOVAL: Unexpected tool selected: '${tool}'."
    }

    emit:
    h5ad = ch_h5ad

    versions = ch_versions
}
