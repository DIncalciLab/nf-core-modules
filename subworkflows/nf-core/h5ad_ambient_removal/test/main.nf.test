nextflow_workflow {
    name 'Test Subworkflow H5AD_AMBIENT_REMOVAL'
    script '../main.nf'
    workflow 'H5AD_AMBIENT_REMOVAL'

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/h5ad_ambient_removal"
    tag "celda/decontx"
    tag "scvitools/scar"
    tag "cellbender/removebackground"
    tag "cellbender/merge"
    tag "soupx"

    test("none") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix.h5ad", checkIfExists: true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_raw_matrix.h5ad", checkIfExists: true)
                ])
                input[1] = 'none'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                {assert snapshot(workflow.out.h5ad).match("res_none_h5ad")},
                {assert snapshot(workflow.out.versions).match("res_none_versions")}
            )
        }
    }

    test("cellbender") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix.h5ad", checkIfExists: true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_raw_matrix.h5ad", checkIfExists: true)
                ])
                input[1] = 'cellbender'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                {assert snapshot(workflow.out.h5ad).match("res_cellbender_h5ad")},
                {assert snapshot(workflow.out.versions).match("res_cellbender_versions")}
            )
        }
    }

    test("decontx") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix.h5ad", checkIfExists: true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_raw_matrix.h5ad", checkIfExists: true)
                ])
                input[1] = 'decontx'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                {assert snapshot(workflow.out.h5ad).match("res_decontx_h5ad")},
                {assert snapshot(workflow.out.versions).match("res_decontx_versions")}
            )
        }
    }

    test("scar") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix.h5ad", checkIfExists: true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_raw_matrix.h5ad", checkIfExists: true)
                ])
                input[1] = 'scar'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                {assert snapshot(workflow.out.h5ad).match("res_scar_h5ad")},
                {assert snapshot(workflow.out.versions).match("res_scar_versions")}
            )
        }
    }

    test("soupx") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test' ], // meta map
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_custom_emptydrops_filter_matrix.h5ad", checkIfExists: true),
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/scdownstream/samples/SAMN14430799_raw_matrix.h5ad", checkIfExists: true)
                ])
                input[1] = 'soupx'
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                {assert snapshot(workflow.out.h5ad).match("res_soupx_h5ad")},
                {assert snapshot(workflow.out.versions).match("res_soupx_versions")}
            )
        }
    }
}