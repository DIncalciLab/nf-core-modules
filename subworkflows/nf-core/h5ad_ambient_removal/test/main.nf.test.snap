{
    "res_none_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "SAMN14430799_custom_emptydrops_filter_matrix_5k.h5ad:md5,0e596860154b0f4529d68219a7f84dc8"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:26:56.259629648"
    },
    "res_scar_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,7e41b26100e3a53ecdaa426111a4d43c"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:45:26.013569479"
    },
    "res_soupx_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,9bd10d702d14d45319fe59f0f593d3a8"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:45:56.172721701"
    },
    "res_cellbender_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,3b7dd429e4f45b698190f2f4df5ad149"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:42:19.496174211"
    },
    "res_decontx_h5ad": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.h5ad:md5,6eef4a7f422b1f7dd7b703c24eebadef"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:42:43.588562279"
    },
    "res_scar_versions": {
        "content": [
            [
                "versions.yml:md5,a3531f350fef8a4704007324ad9fd942"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:45:26.431095774"
    },
    "res_cellbender_versions": {
        "content": [
            [
                "versions.yml:md5,2c66ac05eee7711066b1ffc3228db36e",
                "versions.yml:md5,88b58b33730421b2cdcbe88a72b19953"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:42:19.531658466"
    },
    "res_none_versions": {
        "content": [
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:26:56.301054801"
    },
    "res_decontx_versions": {
        "content": [
            [
                "versions.yml:md5,70bf8966ad13fea31521ab54ec5608cb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:42:43.657887411"
    },
    "res_soupx_versions": {
        "content": [
            [
                "versions.yml:md5,01dcc1e8fd9f08049ecf9db273284213"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-08-05T20:45:56.591887651"
    }
}